tools/migration-unified-annotation-versioning/migration.py

```diff
@@ -0,0 +1,529 @@
+import psycopg2
+import psycopg2.extras
+from psycopg2.extras import RealDictRow
+import math
+import logging
+import datetime
+from pathlib import Path
+import time
+from typing import Dict, Tuple, List, Optional, Callable, Set
+from rich.progress import track
+import msgspec
+import concurrent.futures
+import threading
+from functools import partial
+import heapq
+
+import fossildbapi_pb2 as proto
+import VolumeTracing_pb2 as Volume
+import SkeletonTracing_pb2 as Skeleton
+import Annotation_pb2 as AnnotationProto
+from utils import log_since, batch_range, format_duration, time_str
+from connections import connect_to_fossildb, connect_to_postgres, assert_grpc_success
+
+logger = logging.getLogger("migration-logs")
+checkpoint_logger = logging.getLogger("migration-checkpoints")
+
+
+LayerVersionMapping = Dict[str, Dict[int, int]] # tracing id to (old version to new version)
+MappingIdMap = Dict[str, str] # tracing id to editable mapping id
+
+
+class Migration:
+
+    def __init__(self, args):
+        logger.info(f"Initializing migration with args {args} ...")
+        self.args = args
+        self.src_stub = connect_to_fossildb(args.src, "source")
+        self.dst_stub = None
+        if not args.dry:
+            self.dst_stub = connect_to_fossildb(args.dst, "destination")
+        self.done_count = None
+        self.done_count_lock = threading.Lock()
+        self.failure_count = 0
+        self.failure_count_lock = threading.Lock()
+        self.total_count = None
+        self.before = 0
+
+    def run(self):
+        self.before = time.time()
+        annotations = self.read_annotation_list()
+        self.setup_checkpoint_logging()
+        self.done_count = 0
+        self.failure_count = 0
+        self.total_count = len(annotations)
+
+        with concurrent.futures.ThreadPoolExecutor(max_workers=self.args.num_threads) as executor:
```

Comment:
I think it is no longer, once a couple of threads are used. However, two threads is a lot faster than one. With more threads, the effect fades away. I guess this is because the waits and the CPU-intensive parts like parsing JSON alternate. So if we always have some threads waiting on I/O, the others can do their CPU work in the meantime.

Reply:
Ok. I was just asking because you're using threading and not multiprocessing.

Reply:
That was less a performance consideration and more about work distribution/communication. With the executor I didn’t have to decide in advance which thread takes what work (since the distribution of the annotation sizes is very uneven). With the thread locking I can also print continuous progress and track failures. I guess this can also all be done with multiprocessing, but I’m not very familiar with that. Do you think that would have benefits?

Reply:
If the Python code is the bottleneck that would make sense. For this, multiprocessing would be pretty much a drop-in replacement.